imagenet/main.py

```diff
@@ -29,32 +27,34 @@ def __init__(self, root: pathlib.Path, *, split: str):
         datapipe = dp.iter.LoadFilesFromDisk(datapipe)
         datapipe = dp.iter.ReadFilesFromTar(datapipe)
 
-        (_, stream), datapipe = find(
-            datapipe,
-            "ILSVRC2012_validation_ground_truth.txt",
-            lambda data: pathlib.Path(data[0]).name,
-        )
-        self._val_labels = {
-            f"ILSVRC2012_val_{idx:08d}.JPEG": label
-            for idx, label in enumerate(
-                (int(line.decode().strip()) for line in stream), 1
-            )
-        }
-
-        (_, stream), datapipe = find(
-            datapipe, "meta.mat", lambda data: pathlib.Path(data[0]).name
-        )
-        synsets = scipy.io.loadmat(stream, squeeze_me=True)["synsets"]
-        labels, wnids, clss = zip(
-            *(
-                (label, wnid, tuple(classes.split(", ")))
-                for label, wnid, classes, _, num_children, *_ in synsets
-                if num_children == 0
-            )
-        )
-        self._wnid_to_label = dict(zip(wnids, labels))
-        self._label_to_wnid = dict(zip(labels, wnids))
-        self._label_to_cls = dict(zip(labels, clss))
+        for path, stream in datapipe:
```

Comment:
Indeed this structure has a better performance than what I proposed, but it is harder to maintain. We probably should find a compromise.

This is ok here, since we are only looking for 2 files. Imagine this number grows. We would get one massive `if` / `elif` with an opaque exit condition and exceptions

Reply:
Yep, you are right. For `ImageNet`, I agree with you that it's reasonable to use the `find` function to get code more organized because the meta data is stored in a separate file, which is relatively tiny. So, the overhead can be minimum in this case.